### PR TITLE
[Metadata/IRGen] NFC: Couple of field metadata related improvements

### PR DESCRIPTION
**Pull request by xedin**
This PR does following:

- Enables field descriptor caching (after my previous changes remove field name/type accessors);
- Adjusts structure of the class context descriptor in IRGen

Resolves: rdar://problem/26060144

### CODE REVIEW

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test source compatibility

File: stdlib/public/runtime/MetadataLookup.cpp [hunk 1]
```diff
@@ -425,19 +426,20 @@ _findProtocolDescriptor(llvm::StringRef mangledName) {
...
+  FieldDescriptorCacheEntry(const Metadata *type,
+                            const FieldDescriptor *description)
+      : Type(type), Description(description) {}
 
-  const T *getDescription() { return Description; }
+  const FieldDescriptor *getDescription() { return Description; }
 
   int compareWithKey(const Metadata *other) const {
-    return Base == other;
+    return Type == other ? 0 : (Type < other ? -1 : 1);
```
Comment (jckarter) [inline stdlib/public/runtime/MetadataLookup.cpp]: Comparing unrelated pointers is UB in C/C++. Strictly, you need to do `(uintptr_t)Type < (uintptr_t)Other` there.

**Review by jckarter — CHANGES_REQUESTED** [review]
Just one C semantics nitpick. This LGTM otherwise.

File: stdlib/public/runtime/MetadataLookup.cpp (hunk 1 above)
Comment (xedin) [reply stdlib/public/runtime/MetadataLookup.cpp]: Ah interesting, thanks!

Comment (xedin) [reply stdlib/public/runtime/MetadataLookup.cpp]: Fixed!

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test source compatibility

**Comment by xedin** [comment]
@swift-ci please test source compatibility

**Review by jckarter — APPROVED** [review]

**Pull request merged**